### PR TITLE
[Lang] Support to destory a Snode tree and release memory

### PR DESCRIPTION
**Pull request by beginlner**
Related issue = #2506
API: `<FieldsBuilder>.destroy()`
To achieve it, I allocate a separate memory for a large size SNode tree. We can set the critical size manually by `ti.init(memory_allocate_critical_size=x)`.

### CODE REVIEW

**Comment by beginlner** [comment]
\format

File: python/taichi/snode/fields_builder.py [hunk 1]
```diff
@@ -139,3 +139,8 @@ def finalize(self):
     def _check_not_finalized(self):
         if self._finalized:
             raise InvalidOperationError('FieldsBuilder finalized')
+
+    def destroy(self):
```
Comment (k-ye) [inline python/taichi/snode/fields_builder.py:143]: Consider adding this method to `SNodeTree` itself: The builder is for building the tree. What's the semantics of calling `destroy()` while it's not finalized?

File: taichi/program/program.cpp [hunk 2]
```diff
@@ -509,6 +510,10 @@ void Program::materialize_snode_tree(SNodeTree *tree) {
   }
 }
 
+void Program::destroy_snode_tree(const int snode_tree_id) {
+  memory_pool->destroy_snode_tree(snode_tree_id);
```
Comment (k-ye) [inline taichi/program/program.cpp:514]: This is not the correct level of abstraction. Memory pool should only care about memory allocation. I'd suggest to have a  separate class to manage SNodeTree's memory.

**Review by k-ye — COMMENTED** [review]
Thanks for the quick work! I think the abstraction level needs some improvenment. E.g. you can have a dedicated `SNodeTreeBufferManager` that uses `MemoryPool` and manages the allocated buffers. Let's keep the classes single responsibility.

File: taichi/program/program.cpp (hunk 2 above)
Comment (beginlner) [reply taichi/program/program.cpp:514]: Thanks for your suggestions. I'll rewrite this part of code.

**Comment by ljcc0930** [comment]
Moved to #2564

**Pull request closed without merging**